### PR TITLE
Parameter concat

### LINKED ISSUE #169: [Feature Request] Parameter scheduling
Providing an abstraction to adjust optimizer parameters during training seems like it might be useful - techniques like [SGDR](https://arxiv.org/pdf/1608.03983.pdf) seem applicable to many types of models.

The [`torch.optim.lr_scheduler`](https://pytorch.org/docs/master/optim.html#how-to-adjust-learning-rate) module in PyTorch core implements some useful schedulers, but (a) can only adjust the LR, and (b) only adjusts it per-epoch.

On the other hand, the `Engine` event API seems like a really natural way to adjust parameter values, since handlers that manipulate them could be added for either `ITERATION_*` or `EPOCH_*` events, and modifying multiple parameters at once (e.g. LR and momentum) would be straightforward too.

I wrote a short [IPython notebook](https://gist.github.com/sean-adler/467321a23805736f8dc06a19157b0567) as a prototype of one way it could look to do this with the event API in a general way (plots are at the very bottom). I left most of the actual scheduler code in separate files for now to try and see if the idea is even worth it first. Would this be useful?

### PR DESCRIPTION
**Pull request by amitibo**
Fixes #169 

Description:
Adds ConcatScheduler for concatenating multiple parameter schedulers.

Check list:
* [X] New tests are added (if a new feature is modified)
* [X] New doc strings: text and/or example code are in RST format
* [X] Documentation is updated (if required)

### CODE REVIEW

**Comment by amitibo** [comment]
@vfdev-5 I noticed that the doc description description of *CosineAnnealingScheduler* might be misleading. The parameter is actually annealed from end-value to start-value and not as described, i.e. if one sets start_

**Comment by vfdev-5** [comment]
@amitibo yes you are right, this is even can be observed in the test: 
https://github.com/pytorch/ignite/blob/5432aac9268d1d61de1fd6c956d543e8f07f5bf4/tests/ignite/contrib/handlers/test_param_scheduler.py#L60

https://github.com/pytorch/ignite/blob/5432aac9268d1d61de1fd6c956d543e8f07f5bf4/tests/ignite/contrib/handlers/test_param_scheduler.py#L71-L75

Could you please change the behaviour of that to correspond the common logic ?

**Comment by vfdev-5** [comment]
@amitibo i was thinking more about the names of these classes,would not be better to rename staet, end values by min, max values ?

**Comment by amitibo** [comment]
@vfdev-5 I agree but wouldn't that be a breaking change for people already using these classes?

**Comment by vfdev-5** [comment]
@amitibo actually these classes are not in the stable version 0.1.0 and not in `core`, but in `contrib`. I think we can change this.

**Comment by amitibo** [comment]
@vfdev-5, on a second thought, I am not sure it is correct to rename the start/end_value arguments to min/max_value. In the case of CyclicScheduler, it is valid that the start value of the cycle be either the max or min value along the cycle. The other classes inherit from the CyclicScheduler class, therefore it might be confusing to rename their constructor arguments.

File: ignite/contrib/handlers/param_scheduler.py [hunk 1]
```diff
@@ -64,25 +87,167 @@ def __call__(self, engine):
...
 
 
 class CosineAnnealingScheduler(CyclicalScheduler):
+    """Anneals 'start_value' to 'end_value' over each cycle.
+
+    Args:
+        optimizer (`torch.optim.Optimizer`): the optimizer to use
+        param_name (str): name of optimizer's parameter to update
+        start_value (float): value at start of cycle
+        end_value (float) : value at the end of the cycle
```
Comment (vfdev-5) [inline ignite/contrib/handlers/param_scheduler.py]: @amitibo should not be "value at the **midle** of the cycle" ?

**Comment by vfdev-5** [comment]
@amitibo however the description as 
```
end_value (float) : value at the middle of the cycle
```
is rather confusing.

**Comment by amitibo** [comment]
@vfdev-5 a possible solution is to leave out these arguments from CyclicScheduler and have each subclass define his own set of arguments.

**Comment by amitibo** [comment]
@vfdev-5 I didn't get your point, can you please elaborate?

**Comment by vfdev-5** [comment]
@amitibo I just wanted to say that finally we can keep start/end values instead of min/max values. Another motivation to support start/end names is that we can schedule also the increase of the value and its configuration does not lead to a weird stuff like `max_value=0.1` and `min_value=0.5`

**Comment by vfdev-5** [comment]
I had a thought also to provide a capability as in fast.ai, a sort of value simulation to return the list of values that will be used to update the optimizer. What do you think ?

**Comment by amitibo** [comment]
@vfdev-5 can you please elaborate on your last suggestion? also, can you please mention what is needed for closing this PR?

**Comment by vfdev-5** [comment]
@amitibo the last suggestion can be done in another PR. To close this one it remains the docstring improvement of `CosineAnnealingScheduler`, it is not correct and clear what are `start` and especially `end` values ...

**Comment by amitibo** [comment]
Are you referring to `find_lr()` in fast.ai?

File: ignite/contrib/handlers/param_scheduler.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (amitibo) [reply ignite/contrib/handlers/param_scheduler.py]: In the case of the `CosineAnnealingScheduler` it is actually correct. It is the value at the end f the cycle.

**Comment by vfdev-5** [comment]
No, I thought they had a function to plot lr schedule. Now I found only [this](https://github.com/fastai/fastai/blob/bacffadbd654266af1ba38865f439e0fb6e6c29c/fastai/basic_train.py#L291). But idea is to be able to provide the list of learning rate that will be applied during the training, such that user is sure of what will happend

**Comment by amitibo** [comment]
In my code I have something I call `Parameter` which is more general as it is not limited to the `optimizer` parameters. For example it can be be used to control the weights of multiple losses or to change the network graph. I have a handler that draws the parameters as a `visdom` plot. I am not sure though how to integrate this into `ignite`.

File: ignite/contrib/handlers/param_scheduler.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (vfdev-5) [reply ignite/contrib/handlers/param_scheduler.py]: Yes, I reexecuted my tests, it's true that `end_value` is the value of the end of **demi**-cosine cycle.

```
import numpy as np
import matplotlib.pylab as plt
%matplotlib inline

values = []

cycle_size = 10
start_value = 1.0
end_value = 0.1

for event_index in range(20):
    cycle_progress = event_index / cycle_size
    values.append(start_value + ((end_value - start_value) / 2) * (1 - np.cos(np.pi * cycle_progress)))

plt.plot(values)
```

image


We need to reflect this in the documentation, as ordinary user will think as me that cosine cycle is complete 2*pi cycle.

**Comment by vfdev-5** [comment]
@jasonkriss @alykhantejani please review and merge if OK

**Comment by jasonkriss** [comment]
👍 LGTM!

**Comment by jasonkriss** [comment]
Thanks @amitibo!
